### PR TITLE
Add handling of absd() in Bounds.cpp

### PR DESCRIPTION
**Pull request by steven-johnson**
Formerly this defaulted to bounds-of-type, but that meant that widened expressions never attempted to use the known bounds of the narrow types (e.g. absd(i16(a), i16(b)) would assume [0,65535] even if a and b are known to be i8)

Also, drive-by addition of some debugging infrastructure for Bounds, disabled by default (ala EXPR_MUTATIONS in Simplify)

### CODE REVIEW

File: src/Bounds.cpp [hunk 1]
```diff
@@ -861,6 +965,44 @@ class Bounds : public IRVisitor {
...
+                Expr e = abs(op->args[0] - op->args[1]);
+                e.accept(this);
+            } else {
+                // absd() for int types will always produce a uint result
+                internal_assert(t.is_uint());
+
+                // Recover what the matched types for the args were
+                Expr a = op->args[0];
+                Expr b = op->args[1];
+                match_types(a, b);
```
Comment (abadams) [inline src/Bounds.cpp]: Why is this necessary if the absd function already did it?

File: src/Bounds.cpp [hunk 2]
```diff
@@ -861,6 +965,44 @@ class Bounds : public IRVisitor {
...
+                b.accept(this);
+                Interval b_interval = interval;
+
+                a_interval.min = simplify(a_interval.min);
+                a_interval.max = simplify(a_interval.max);
+                b_interval.min = simplify(b_interval.min);
+                b_interval.max = simplify(b_interval.max);
+
+                if (a_interval.is_bounded() && b_interval.is_bounded()) {
+                    // Cast to 64-bit type to minimize cast-out-of-range issues;
```
Comment (abadams) [inline src/Bounds.cpp]: To avoid introducing wider types, how about:

interval.max = max(absd(a_interval.max, b_interval.min), absd(a_interval.min, b_interval.max))

File: src/Bounds.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/Bounds.cpp]: You are correct, the match_type'd versions of the args are saved in the op.

File: src/Bounds.cpp (hunk 2 above)
Comment (steven-johnson) [reply src/Bounds.cpp]: ah yes, but brings up another fun fact: `absd()` isn't handled in Simplify_Call.cpp, so `max(absd((uint8)20, (uint8)5), absd((uint8)10, (uint8)30))` isn't simplified down to `(uint8)20` as it should be, thus the newly added test fails

Comment (steven-johnson) [reply src/Bounds.cpp]: (adding absd() handling to simplify now)
